Source/Dafny/Dafny.atg

```diff
@@ -1819,16 +1913,11 @@ TypeAndToken<out IToken tok, out Type ty, bool inExpressionContext>
                                     .)
   | arrayToken                      (. tok = t; .)
     OptGenericInstantiation<out gt, inExpressionContext>
-                                    (. var dimString = tok.val.Substring(5);
-                                       int dims = StringToInt(dimString, 1, "arrays of that many dimensions");
-                                       ty = theBuiltIns.ArrayType(tok, dims, gt, true);
-                                    .)
-  | arrayToken_q                    (. tok = t; .)
-    OptGenericInstantiation<out gt, inExpressionContext>
-                                    (. var dimString = tok.val.Substring(5);
-                                       dimString = dimString.Substring(0, dimString.Length - 1);
+                                    (. var tokString = tok.val;
+                                       bool q = tokString[tokString.Length-1] == '?';
+                                       var dimString = tokString.Substring(5, tokString.Length - (q?6:5)); // 5 is length of "array"
```

Comment:
A bit confused why `tokString.Length - (q?6:5)` is necessary here. An extra comment would be helpful.

Reply:
OK - comment added